src/plus/gk/serverConnection.ts

```diff
@@ -169,28 +192,121 @@ export class ServerConnection implements Disposable {
 					url = `${url}?${options.query}`;
 				}
 			}
-			// TODO@eamodio handle common response errors
 
-			return this.fetch(
+			const result = await this.fetch(
 				url,
 				{
 					...init,
 					headers: headers as HeadersInit,
 				},
 				options,
 			);
+			this.resetRequestExceptionCount();
+			return result;
 		} catch (ex) {
-			Logger.error(ex, scope);
+			this.handleRequestError('gitkraken', ex, scope);
```

Comment:
When we use node-fetch, in some cases, including with our backend, it is possible for the fetch call not to throw, but still get an error status. In other words, it is possible for the response to be a 429, and we never reach this `catch` block. In this case, we check `response.ok` and `response.status` to determine if an error occurred, and what code, since the `fetch` call itself will not throw. [Here is an example of this in the ai provider](https://github.com/gitkraken/vscode-gitlens/blob/main/src/ai/openaiProvider.ts#L200-L230).

We still want to track these, and we still want to stop the requests if they fail (i.e. `handleRequestError`) but we may need to be careful about throwing to the caller in these cases, as the caller may already have handling at the higher level. For example `draftsService` has `handleBadDraftResponse` and `cloudIntegrationService` still needs to send telemetry, etc.

Reply:
@axosoft-ramint
Should it also be fixed in GitHub and GitLab providers?

Reply:
@axosoft-ramint 

Something like this? https://github.com/gitkraken/vscode-gitlens/commit/49fc67e895a7f78291b3fc92b37dee5a90394caf
Here I'm wrapping an unsuccessful response to an error in order to handle it along with exceptions.

(and I've locked the error name to fight the type checks in the handler).

Reply:
@sergeibbb We can keep scope reduced to just our fetch calls from now, and update integration queries later if we need to.

The commit you linked seems to be handling it, although we should note now that these errors are being thrown to the caller where they were not before, so all the services that have handling for `!response.ok` from fetch results that route through `serverConnection` should now be updated to `try/catch` and we should ensure they are still handling the same errors appropriately in the `catch` blocks, and using `finally` as needed.

Reply:
@axosoft-ramint 

Well, I've completed the investigation. This change ends up in too many places:
<img width="660" alt="image" src="https://github.com/user-attachments/assets/5966e028-296e-4d1c-8fb6-eafdd9d27e37">

Therefore I've changed my handling function. I let it detect timeout and rate limit responses and track them. Then it just exist and lets consumers to handle the result as they have been doing it before, so nothing is changed for them. Here is how I'm doing it: https://github.com/gitkraken/vscode-gitlens/commit/85bf10c7ea4333cc67b14a8e7a3d528a3d9124d2